source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst

```diff
@@ -0,0 +1,101 @@
+      import edu.wpi.first.shuffleboard.api.plugin.Plugin;
+
+      @Description(group = "com.example", name = "MyPlugin", version = "1.2.3", summary = "An example plugin")
+      public class MyPlugin extends Plugin {
+
+      }
+
+More resources on how these attributes are used, including version numbers Shuffleboard's plugin dependencies can be found `here <https://semver.org/>`_.
+
+Note the ``@Description`` annotation which is needed to tell the plugin loader the properties of the custom plugin class.
+Plugin classes are permitted to have a defualt constructor but it cannot take any arguments.
+
+Building plugin
+---------------
+Plugins require the usuage of the `Shuffleboard API Library <https://frcmaven.wpi.edu/artifactory/release/edu/wpi/first/shuffleboard/api/>`_. These dependencies can be resolved in the
+``build.gradle`` file or using maven. The dependencies would be as follows:
+
+For Gradle:
+
+.. code-block:: groovy
+
+   repositories {
+      mavenCentral()
+      maven{ url "https://frcmaven.wpi.edu:443/artifactory/release" }
+   }
+
+   dependencies {
+      compileOnly("edu.wpi.first.shuffleboard", "api", "2020.+")
+      compileOnly("edu.wpi.first.shuffleboard.plugin", "networktables", "2020.+")
+   }
+
+For Maven:
+
+ .. code-blocl:: groovy
+
+Plugins are allowed to have dependencies on other plugins and libraries, however, they must be included correctly in the maven or gradle build file.
+When having Plugin dependencies to other plugins, it is good practise to define those dependencies so the plugin does not load when the dependencies do not load as well.
+This can be done using the ``@Requires`` annotation as shown below:
+
+.. code-block:: java
+
+   @Requires(group = "com.example", name = "Good Plugin", minVersion = "1.2.3")
+   @Requires(group = "edu.wpi.first.shuffleboard", "Base", minVersion = "1.0.0")
+   @Description(group = "com.example", name = "MyPlugin", version = "1.2.3", summary = "An example plugin")
+   public class MyPlugin extends Plugin {
+
+   }
+
+The ``minVersion`` specifies the minimum allowable version of the plugin that can be loaded. For example, if the ``minVersion`` is 1.4.5, and the plugin with the version 1.4.7 is loaded, it will be allowed to do so. However, if the plugin with the version 1.2.4 is loaded, it will not be allowed to since it is less than the ``minVersion``.
+
+Deploying Plugin To Shuffleboard
+--------------------------------
+In order to load a plugin to the shuffleboard, you will need to generate a jar file of the plugin and put it in the ``~/Shuffleboard/plugins`` folder. This can be done automatically
+from gradle as noted:
+
+.. code-block::groovy
+
+   task deployPlugin (type: Copy, group: "...", description: "...", dependsOn: "build") {
+      from "build/libs"
+      into "path/to/Shuffleboard/plugins"
+      include "*.jar"
+   }
+
+The path to your Shuffleboard plugin folder will most likely be ``~/Shuffleboard/plugins`` or ``~/wpilib/Shuffleboard/plugins`` .
```

Comment:
I do not recommend mentioning ~/wpilib/Shuffleboard/plugins until it is changed in Shuffleboard